### PR TITLE
lazily cached serializer identifier

### PR DESCRIPTION
**Pull request by Aaronontheweb**
This improved the performance of Akka.Remote by 2.5x and reduced gen0 garbage collection by 33%. Found it while profiling RemotePingPong on another branch.

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
Hmm, drat

```
Cause: System.Runtime.Serialization.SerializationException: Cannot find manifest class [null] for serializer with id [0].
   at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest)
   at Akka.Remote.Serialization.DaemonMsgCreateSerializer.PropsFromProto(PropsData protoProps)
   at Akka.Remote.Serialization.DaemonMsgCreateSerializer.FromBinary(Byte[] bytes, Type type)
   at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest)
   at Akka.Remote.DefaultMessageDispatcher.Dispatch(IInternalActorRef recipient, Address recipientAddress, Payload message, IActorRef senderOption)
   at Akka.Remote.EndpointReader.<Reading>b__11_1(InboundPayload inbound)
   at lambda_method(Closure , Object , Action`1 , Action`1 , Action`1 )
   at Akka.Tools.MatchHandler.PartialHandlerArgumentsCapture`4.Handle(T value)
   at Akka.Actor.ReceiveActor.ExecutePartialMessageHandler(Object message, PartialAction`1 partialAction)
   at Akka.Actor.UntypedActor.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Akka.Actor.ActorCell.HandleFailed(Failed f)
   at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)
```

Looks like I'll need to add a null check - I guess `Type.Get` handles this under the hood, Our cache will need to also.

**Comment by Aaronontheweb** [comment]
(that's what is failing the remote deployment MNTR spec consistently)

**Comment by Aaronontheweb** [comment]
Mono build has hung a couple of times also. Might be related.

File: src/core/Akka/Util/Reflection/TypeCache.cs [hunk 1]
```diff
@@ -15,7 +16,7 @@ namespace Akka.Util.Reflection
     /// </summary>
     public static class TypeCache
     {
-        private static readonly ConcurrentDictionary<string, Type> TypeMap = new ConcurrentDictionary<string, Type>();
+        private static readonly ConcurrentDictionary<string, Type> TypeMap = new ConcurrentDictionary<string, Type>(new[] { new KeyValuePair<string, Type>("null", null),  });
```
Comment (Aaronontheweb) [inline src/core/Akka/Util/Reflection/TypeCache.cs]: This is a hack, but it avoids having to add a `str.Equals("null")` check for literally every type.

File: src/core/Akka/Serialization/Serializer.cs [hunk 2]
```diff
@@ -39,20 +39,23 @@ public abstract class Serializer
...
             this.system = system;
+            _value = new FastLazy<int>(() => SerializerIdentifierHelper.GetSerializerIdentifierFromConfig(GetType(), system));
         }
 
         /// <summary>
         /// Completely unique value to identify this implementation of Serializer, used to optimize network traffic
         /// Values from 0 to 16 is reserved for Akka internal usage
         /// </summary>
-        public virtual int Identifier => SerializerIdentifierHelper.GetSerializerIdentifierFromConfig(GetType(), system);
+        public virtual int Identifier => _value.Value;
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serializer.cs]: You could change it to `Identifier  { get; }  =... ` and drop FastLazy

**Pull request merged**